File: android/src/main/java/expo/modules/applemusic/LibraryRestClient.kt
package expo.modules.applemusic

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject

/** Library-domain Apple Music REST (user collection reads + playback id resolution). */
internal class LibraryRestClient(
  private val transport: AppleMusicRestTransport,
) {
  suspend fun getLibraryPlaylists(musicUserToken: String, limit: Int, offset: Int): List<Map<String, Any?>> =
    withContext(Dispatchers.IO) {
      val json =
        transport.getJson(
          musicUserToken,
          "/v1/me/library/playlists",
          mapOf("limit" to limit.toString(), "offset" to offset.toString()),
        )
      mapTopLevelResourceArray(json) { AppleMusicJsonMapper.mapPlaylist(it) }
    }

  suspend fun getLibrarySongs(musicUserToken: String, limit: Int, offset: Int): List<Map<String, Any?>> =
    withContext(Dispatchers.IO) {
      val json =
        transport.getJson(
          musicUserToken,
          "/v1/me/library/songs",
          mapOf("limit" to limit.toString(), "offset" to offset.toString()),
        )
      mapTopLevelResourceArray(json) { AppleMusicJsonMapper.mapSong(it) }
    }

  suspend fun getPlaylistTracks(musicUserToken: String, playlistId: String): List<Map<String, Any?>> =
    withContext(Dispatchers.IO) {
      val json = transport.getJson(
          musicUserToken,
          "/v1/me/library/playlists/$playlistId/tracks")
      val data = requireDataArray(json)
      buildList {
        for (i in 0 until data.length()) {
          val resource = data.getJSONObject(i)
          if (resource.optString("type", "").contains("song")) {
            add(AppleMusicJsonMapper.mapSong(resource))
          }
        }
      }
    }

  private fun librarySearchTypeParam(type: String): String? =
    when (type) {
      "library-songs", "songs" -> "library-songs"
      "library-albums", "albums" -> "library-albums"
      "library-artists", "artists" -> "library-artists"
      "library-playlists", "playlists" -> "library-playlists"
      "library-music-videos", "music-videos", "musicVideos" -> "library-music-videos"
      else -> null
    }

  suspend fun getLibraryArtists(musicUserToken: String, limit: Int, offset: Int): List<Map<String, Any?>> =
    withContext(Dispatchers.IO) {
      val json =
        transport.getJson(
          musicUserToken,
          "/v1/me/library/artists",
          mapOf("limit" to limit.toString(), "offset" to offset.toString()),
        )
      mapTopLevelResourceArray(json) { AppleMusicJsonMapper.mapArtist(it) }
    }

  suspend fun getLibraryAlbums(musicUserToken: String, limit: Int, offset: Int): List<Map<String, Any?>> =
    withContext(Dispatchers.IO) {
      val json =
        transport.getJson(
          musicUserToken,
          "/v1/me/library/albums",
          mapOf("limit" to limit.toString(), "offset" to offset.toString()),
        )
      mapTopLevelResourceArray(json) { AppleMusicJsonMapper.mapAlbum(it) }
    }

  data class LibrarySearchResult(
    val songs: List<Map<String, Any?>>,
    val albums: List<Map<String, Any?>>,
    val artists: List<Map<String, Any?>>,
    val playlists: List<Map<String, Any?>>,
    val musicVideos: List<Map<String, Any?>>,
  )

  suspend fun getLibraryMusicVideos(musicUserToken: String, limit: Int, offset: Int): List<Map<String, Any?>> =
    withContext(Dispatchers.IO) {
      val json =
        transport.getJson(
          musicUserToken,
          "/v1/me/library/music-videos",
          mapOf("limit" to limit.toString(), "offset" to offset.toString()),
        )
      mapTopLevelResourceArray(json) { AppleMusicJsonMapper.mapMusicVideo(it) }
    }

  suspend fun searchLibrary(
    musicUserToken: String,
    term: String,
    types: List<String>,
    limit: Int,
    offset: Int,
  ): LibrarySearchResult =
    withContext(Dispatchers.IO) {
      val typeParam =
        types
          .mapNotNull { librarySearchTypeParam(it) }
          .distinct()
          .sorted()
          .joinToString(",")
          .ifEmpty { "library-songs,library-albums" }

      val json =
        transport.getJson(
          musicUserToken,
          "/v1/me/library/search",
          mapOf(
            "term" to term,
            "types" to typeParam,
            "limit" to limit.toString(),
            "offset" to offset.toString(),
          ),
        )

      val results = json.optJSONObject("results") ?: JSONObject()
      LibrarySearchResult(
        songs =
          mapResourceArray(results.optJSONObject("library-songs")?.optJSONArray("data")) {
            AppleMusicJsonMapper.mapSong(it)
          },
        albums =
          mapResourceArray(results.optJSONObject("library-albums")?.optJSONArray("data")) {
            AppleMusicJsonMapper.mapAlbum(it)
          },
        artists =
          mapResourceArray(results.optJSONObject("library-artists")?.optJSONArray("data")) {
            AppleMusicJsonMapper.mapArtist(it)
          },
        playlists =
          mapResourceArray(results.optJSONObject("library-playlists")?.optJSONArray("data")) {
            AppleMusicJsonMapper.mapPlaylist(it)
          },
        musicVideos =
          mapResourceArray(results.optJSONObject("library-music-videos")?.optJSONArray("data")) {
            AppleMusicJsonMapper.mapMusicVideo(it)
          },
      )
    }

  /**
   * Lightweight library probe for subscription inference.
   * Returns `false` only on HTTP 403 / `permissionDenied` (no library / subscription).
   * Network failures, missing tokens, and other API errors are rethrown.
   */
  suspend fun probeLibraryAccess(musicUserToken: String): Boolean =
    withContext(Dispatchers.IO) {
      try {
        transport.getJson(
          musicUserToken,
          "/v1/me/library/songs", mapOf("limit" to "1"))
        true
      } catch (error: expo.modules.kotlin.exception.CodedException) {
        if (
          error.code == AppleMusicErrorCodes.PERMISSION_DENIED &&
          error.message?.contains("403") == true
        ) {
          false
        } else {
          throw error
        }
      }
    }

  suspend fun resolveCatalogPlaybackId(musicUserToken: String, libraryId: String, mediaType: String): String =
    withContext(Dispatchers.IO) {
      val path =
        when (mediaType) {
          "song" -> "/v1/me/library/songs/$libraryId"
          "album" -> "/v1/me/library/albums/$libraryId"
          "playlist" -> "/v1/me/library/playlists/$libraryId"
          else -> throw AppleMusicErrors.unknownMediaType(mediaType)
        }
      val json = transport.getJson(musicUserToken, path)
      val resource = json.getJSONArray("data").getJSONObject(0)
      AppleMusicJsonMapper.catalogPlaybackId(resource)
        ?: throw AppleMusicErrors.itemNotFound(mediaType.replaceFirstChar { it.uppercase() }, true)
    }

  suspend fun resolveLibrarySongCatalogIds(musicUserToken: String, playlistId: String): List<String> =
    withContext(Dispatchers.IO) {
      val json = transport.getJson(
          musicUserToken,
          "/v1/me/library/playlists/$playlistId/tracks")
      val data = requireDataArray(json)
      buildList {
        for (i in 0 until data.length()) {
          val resource = data.getJSONObject(i)
          if (!resource.optString("type", "").contains("song")) continue
          AppleMusicJsonMapper.catalogPlaybackId(resource)?.let { add(it) }
        }
      }
    }
}

File: src/rest/library-rest-client.ts
import {
  catalogPlaybackId,
  mapAlbum,
  mapArtist,
  mapMusicVideo,
  mapPlaylist,
  mapSong,
  type AppleMusicApiResource,
} from '../mappers/apple-music-json-mapper';
import { AppleMusicErrorCode } from '../constants/apple-music-error-codes';
import type { LibrarySearchType } from '../types/library-search';
import { isAppleMusicError } from '../utils/apple-music-error';
import * as errors from '../web/apple-music-errors';
import type { AppleMusicRestTransport } from './apple-music-rest-transport';
import { mapResourceArray, mapTopLevelResourceArray, parseDataArray } from './rest-json';

export type LibrarySearchResult = {
  songs: Record<string, unknown>[];
  albums: Record<string, unknown>[];
  artists: Record<string, unknown>[];
  playlists: Record<string, unknown>[];
  musicVideos: Record<string, unknown>[];
};

function librarySearchTypeParam(type: string): string | null {
  switch (type) {
    case 'library-songs':
    case 'songs':
      return 'library-songs';
    case 'library-albums':
    case 'albums':
      return 'library-albums';
    case 'library-artists':
    case 'artists':
      return 'library-artists';
    case 'library-playlists':
    case 'playlists':
      return 'library-playlists';
    case 'library-music-videos':
    case 'music-videos':
    case 'musicVideos':
      return 'library-music-videos';
    default:
      return null;
  }
}

/** Library-domain Apple Music REST (user collection reads + playback id resolution). */
export class LibraryRestClient {
  constructor(private readonly transport: AppleMusicRestTransport) {}

  async getLibraryPlaylists(musicUserToken: string, limit: number, offset: number) {
    const json = await this.transport.getJson('/v1/me/library/playlists', {
      limit: String(limit),
      offset: String(offset),
    }, musicUserToken);
    return mapTopLevelResourceArray(json.data, mapPlaylist);
  }

  async getLibrarySongs(musicUserToken: string, limit: number, offset: number) {
    const json = await this.transport.getJson('/v1/me/library/songs', {
      limit: String(limit),
      offset: String(offset),
    }, musicUserToken);
    return mapTopLevelResourceArray(json.data, mapSong);
  }

  async getPlaylistTracks(musicUserToken: string, playlistId: string) {
    const json = await this.transport.getJson(
      `/v1/me/library/playlists/${playlistId}/tracks`,
      {},
      musicUserToken,
    );
    const data = parseDataArray(json.data);
    return data
      .filter((item) => String((item as AppleMusicApiResource).type ?? '').includes('song'))
      .map((item) => mapSong(item as AppleMusicApiResource));
  }

  async getLibraryArtists(musicUserToken: string, limit: number, offset: number) {
    const json = await this.transport.getJson('/v1/me/library/artists', {
      limit: String(limit),
      offset: String(offset),
    }, musicUserToken);
    return mapTopLevelResourceArray(json.data, mapArtist);
  }

  async getLibraryAlbums(musicUserToken: string, limit: number, offset: number) {
    const json = await this.transport.getJson('/v1/me/library/albums', {
      limit: String(limit),
      offset: String(offset),
    }, musicUserToken);
    return mapTopLevelResourceArray(json.data, mapAlbum);
  }

  async getLibraryMusicVideos(musicUserToken: string, limit: number, offset: number) {
    const json = await this.transport.getJson('/v1/me/library/music-videos', {
      limit: String(limit),
      offset: String(offset),
    }, musicUserToken);
    return mapTopLevelResourceArray(json.data, mapMusicVideo);
  }

  async searchLibrary(
    musicUserToken: string,
    term: string,
    types: LibrarySearchType[],
    limit: number,
    offset: number,
  ): Promise<LibrarySearchResult> {
    const typeParam = [...new Set(types.map(librarySearchTypeParam).filter(Boolean))].sort().join(
      ',',
    );
    const typesQuery = typeParam || 'library-songs,library-albums';

    const json = await this.transport.getJson('/v1/me/library/search', {
      term,
      types: typesQuery,
      limit: String(limit),
      offset: String(offset),
    }, musicUserToken);

    const results = (json.results ?? {}) as Record<string, { data?: AppleMusicApiResource[] }>;
    return {
      songs: mapResourceArray(results['library-songs']?.data, mapSong),
      albums: mapResourceArray(results['library-albums']?.data, mapAlbum),
      artists: mapResourceArray(results['library-artists']?.data, mapArtist),
      playlists: mapResourceArray(results['library-playlists']?.data, mapPlaylist),
      musicVideos: mapResourceArray(results['library-music-videos']?.data, mapMusicVideo),
    };
  }

  /**
   * Lightweight library probe for subscription inference.
   * Returns `false` only on HTTP 403 / subscription-or-permission denial.
   * Network failures, missing tokens, and other API errors are rethrown.
   */
  async probeLibraryAccess(musicUserToken: string): Promise<boolean> {
    try {
      await this.transport.getJson('/v1/me/library/songs', { limit: '1' }, musicUserToken);
      return true;
    } catch (error) {
      if (
        isAppleMusicError(error) &&
        error.code === AppleMusicErrorCode.permissionDenied &&
        error.message.includes('403')
      ) {
        return false;
      }
      throw error;
    }
  }

  async resolveCatalogPlaybackId(musicUserToken: string, libraryId: string, mediaType: string): Promise<string> {
    const path =
      mediaType === 'song'
        ? `/v1/me/library/songs/${libraryId}`
        : mediaType === 'album'
          ? `/v1/me/library/albums/${libraryId}`
          : mediaType === 'playlist'
            ? `/v1/me/library/playlists/${libraryId}`
            : null;
    if (!path) {
      throw errors.unknownMediaType(mediaType);
    }
    const json = await this.transport.getJson(path, {}, musicUserToken);
    const rows = parseDataArray(json.data);
    const data = rows[0];
    if (!data) {
      throw errors.itemNotFound(mediaType, true);
    }
    const catalogId = catalogPlaybackId(data as AppleMusicApiResource);
    if (!catalogId) {
      throw errors.itemNotFound(mediaType, true);
    }
    return catalogId;
  }

  async resolveLibrarySongCatalogIds(musicUserToken: string, playlistId: string): Promise<string[]> {
    const json = await this.transport.getJson(
      `/v1/me/library/playlists/${playlistId}/tracks`,
      {},
      musicUserToken,
    );
    const data = parseDataArray(json.data);
    const ids: string[] = [];
    for (const item of data) {
      const resource = item as AppleMusicApiResource;
      if (!String(resource.type ?? '').includes('song')) {
        continue;
      }
      const catalogId = catalogPlaybackId(resource);
      if (catalogId) {
        ids.push(catalogId);
      }
    }
    return ids;
  }
}
